File: backend/src/services/subsidiaryProjectBlockService.ts
import { AppDataSource } from "../config/data-source";
import { SubsidiaryProjectBlock } from "../entities/SubsidiaryProjectBlock";
import { ConflictError, SubsidiaryProjectBlockedError } from "../utils/errors";

/** Every currently-blocked (subsidiary, project code) pair, newest first —
 * the admin management view. */
export function listSubsidiaryProjectBlocks(): Promise<SubsidiaryProjectBlock[]> {
  return AppDataSource.getRepository(SubsidiaryProjectBlock).find({ order: { createdAt: "DESC" } });
}

/** Blocks a (subsidiary, project code) pair from new uploads — exact-match on
 * both (both come from dropdowns, not free text). Rejects an exact-duplicate
 * pair with a 409 rather than a raw DB unique-constraint error. */
export async function createSubsidiaryProjectBlock(
  subsidiaryName: string,
  projectCode: string,
): Promise<SubsidiaryProjectBlock> {
  const repo = AppDataSource.getRepository(SubsidiaryProjectBlock);
  const existing = await repo.findOne({ where: { subsidiaryName, projectCode } });
  if (existing) {
    throw new ConflictError(`"${projectCode}" is already blocked for subsidiary "${subsidiaryName}"`);
  }
  return repo.save(repo.create({ subsidiaryName, projectCode }));
}

/** Unblocks a pair (deletes the row). Returns false if it wasn't blocked —
 * callers map that to a 404, same convention as the rest of the admin API. */
export async function deleteSubsidiaryProjectBlock(id: string): Promise<boolean> {
  const result = await AppDataSource.getRepository(SubsidiaryProjectBlock).delete({ id });
  return (result.affected ?? 0) > 0;
}

/**
 * Called from uploadService.createUpload before anything is written to disk,
 * alongside and independently of projectCodeService's own open check: throws
 * SubsidiaryProjectBlockedError if an admin has specifically blocked this
 * project code for this subsidiary. A no-op (resolves) otherwise — this is a
 * blocklist, not a picklist, so an unknown pair is simply not blocked.
 */
export async function assertNotBlocked(subsidiaryName: string, projectCode: string): Promise<void> {
  const blocked = await AppDataSource.getRepository(SubsidiaryProjectBlock).findOne({
    where: { subsidiaryName, projectCode },
  });
  if (blocked) {
    throw new SubsidiaryProjectBlockedError(
      `Project code "${projectCode}" is closed for new uploads from subsidiary "${subsidiaryName}"`,
    );
  }
}

File: backend-py/app/services/subsidiary_project_block_service.py
"""Port of `backend/src/services/subsidiaryProjectBlockService.ts`."""

from __future__ import annotations

from sqlalchemy import delete, select
from sqlalchemy.orm import Session

from app.errors import ConflictError, SubsidiaryProjectBlockedError
from app.models.subsidiary_project_block import SubsidiaryProjectBlock


def list_subsidiary_project_blocks(db: Session) -> list[SubsidiaryProjectBlock]:
    """Every currently-blocked (subsidiary, project code) pair, newest
    first — the admin management view."""
    return list(
        db.execute(select(SubsidiaryProjectBlock).order_by(SubsidiaryProjectBlock.createdAt.desc())).scalars().all()
    )


def create_subsidiary_project_block(db: Session, subsidiary_name: str, project_code: str) -> SubsidiaryProjectBlock:
    """Blocks a (subsidiary, project code) pair from new uploads —
    exact-match on both. Rejects an exact-duplicate pair with a
    `ConflictError`."""
    existing = db.execute(
        select(SubsidiaryProjectBlock).where(
            SubsidiaryProjectBlock.subsidiaryName == subsidiary_name,
            SubsidiaryProjectBlock.projectCode == project_code,
        )
    ).scalar_one_or_none()
    if existing is not None:
        raise ConflictError(f'"{project_code}" is already blocked for subsidiary "{subsidiary_name}"')

    created = SubsidiaryProjectBlock(subsidiaryName=subsidiary_name, projectCode=project_code)
    db.add(created)
    db.commit()
    db.refresh(created)
    return created


def delete_subsidiary_project_block(db: Session, id: str) -> bool:
    """Unblocks a pair (deletes the row). Returns `False` if it wasn't
    blocked — callers map that to a 404."""
    result = db.execute(delete(SubsidiaryProjectBlock).where(SubsidiaryProjectBlock.id == id))
    db.commit()
    return (result.rowcount or 0) > 0


def assert_not_blocked(db: Session, subsidiary_name: str, project_code: str) -> None:
    """Raises `SubsidiaryProjectBlockedError` if an admin has specifically
    blocked this project code for this subsidiary. A no-op otherwise — this
    is a blocklist, not a picklist. Not called from any route in this
    phase — ported so the form-builder phase can use it directly."""
    blocked = db.execute(
        select(SubsidiaryProjectBlock).where(
            SubsidiaryProjectBlock.subsidiaryName == subsidiary_name,
            SubsidiaryProjectBlock.projectCode == project_code,
        )
    ).scalar_one_or_none()
    if blocked is not None:
        raise SubsidiaryProjectBlockedError(
            f'Project code "{project_code}" is blocked for subsidiary "{subsidiary_name}"'
        )
